### PR TITLE
Fixes #30, opset 9 updates: PRelu, Flatten, MatMul, BatchNorm, Upsample

### LINKED ISSUE #30: Support for Opset 9
For example, `BatchNormalization` should not have a `spatial` attribute in opset 9. For more examples of implementation, please see https://github.com/onnx/onnxmltools/pull/241.

### PR DESCRIPTION
**Pull request by vinitra**
- ported over common op updates from onnxmltools for opset 9 (batchNorm, prelu, upsample)
- updated flatten op: bidirectional.py
- updated matmul op: dense.py
- ~~reorganized onnx_ops file in alphabetical order~~ (will be addressed in separate PR)
- ~~ported over crop common op [onnxmltools:249](https://github.com/onnx/onnxmltools/pull/249), deprecated from experimental op in opset 10~~  (will be addressed in separate PR)

### CODE REVIEW

File: keras2onnx/common/onnx_ops.py [hunk 1]
```diff
@@ -69,21 +65,17 @@ def _apply_pointwise_operation(scope, op_type, input_names, output_name, contain
...
 
-
 def apply_batch_norm(scope, input_names, output_names, container, operator_name=None,
                      epsilon=None, is_test=None, momentum=None, spatial=None):
     name = _create_name_or_use_existing_one(scope, 'BatchNormalization', operator_name)
-
-    attrs = {'name': name, 'epsilon': epsilon, 'momentum': momentum, 'spatial': spatial}
+    attrs = {'name': name, 'epsilon': epsilon, 'momentum': momentum}
 
     if container.target_opset < 6:
```
Comment (jiafatom) [inline keras2onnx/common/onnx_ops.py]: Can we write like 
`if container.target_opset < 9:      attrs['spatial'] = spatial`
This way we only need to assign attrs['spatial'] once, compared you need write it 3 times here. Same for is_test

File: keras2onnx/common/onnx_ops.py [hunk 2]
```diff
@@ -112,28 +108,46 @@ def apply_cast(scope, input_name, output_name, container, operator_name=None, to
...
+            attrs['to'] = allowed_type_name_and_type_enum_pairs[to]
+            op_version = 1
+        else:
+            # Enum, for example, TensorProto.INT64
+            attrs['to'] = to
+            op_version = 6
     else:
-        # Enum, for example, TensorProto.INT64
+        # Enum value, for example, TensorProto.INT64
+        # String casting is supported in opset 9
```
Comment (jiafatom) [inline keras2onnx/common/onnx_ops.py]: Can we have a unit test for string casting, we only need this unit test for opset 9

File: keras2onnx/common/onnx_ops.py [hunk 3]
```diff
@@ -346,70 +431,13 @@ def apply_upsample(scope, input_name, output_name, container, operator_name=None
         attrs['width_scale'] = float(scales[3])
         attrs['mode'] = mode.upper()
         op_version = 1
-    else:
+    elif container.target_opset < 9:
         attrs['scales'] = list(map(float, scales))
         attrs['mode'] = mode.lower()
```
Comment (jiafatom) [inline keras2onnx/common/onnx_ops.py]: when target_opset == 9, shall we also  attrs['mode'] = mode.lower() ?

File: keras2onnx/common/onnx_ops.py (hunk 1 above)
Comment (vinitra) [reply keras2onnx/common/onnx_ops.py]: Thanks for the feedback, @jiafatom! Resolved in latest update.

File: keras2onnx/common/onnx_ops.py (hunk 3 above)
Comment (vinitra) [reply keras2onnx/common/onnx_ops.py]: Refactored to avoid replicated code in the else case, which encompasses opset 9 and future opsets!

**Comment by wenbingl** [comment]
Is the file now fully as same as the one in onnxmltools? we should try to keep them same so that it would be nice for us for the future merge.

**Comment by jiafatom** [comment]
This PR reorganized onnx_ops file in alphabetical order, can we separate this reorganization into another PR? (We can then proceed the reorganization PR). Then it is easier to check the PR to see what is the real logic change.

File: keras2onnx/ke2onnx/bidirectional.py [hunk 4]
```diff
@@ -253,9 +253,12 @@ def convert_bidirectional(scope, operator, container):
             perm = [1, 0, 2] if container.target_opset <= 5 else [2, 0, 1, 3]
             apply_transpose(scope, lstm_h_name, transposed_h_name, container, perm=perm)
 
+            # Maintain backwards opset compatibility for 'Flatten'
+            target_opset = 1 if container.target_opset < 9 else 9
+
             # Flatten ONNX (N, D, C') into (N, D * C')
             container.add_node('Flatten', transposed_h_name, operator.outputs[0].full_name,
-                               name=scope.get_unique_variable_name('Flatten'), axis=1)
+                               name=scope.get_unique_variable_name('Flatten'), axis=1, op_version=target_opset)
```
Comment (wenbingl) [inline keras2onnx/ke2onnx/bidirectional.py]: Semantically,  op_version here is op.since_version, not the parameter the user specified.

Comment (vinitra) [reply keras2onnx/ke2onnx/bidirectional.py]: addressed by renaming variable to op_version (as consistent with other conversions)

**Comment by vinitra** [comment]
Hi @jiafatom, I separated the alphabetical reordering changes away from the actual content changes. Any suggestions for how to help resolve the failing test in test_layers.py on linux?

File: keras2onnx/common/onnx_ops.py (hunk 2 above)
Comment (vinitra) [reply keras2onnx/common/onnx_ops.py]: Discussed offline, added TODO for unit test as cast operator is not yet supported in Keras->ONNX.

**Comment by wenbingl** [comment]
> Hi @jiafatom, I separated the alphabetical reordering changes away from the actual content changes. Any suggestions for how to help resolve the failing test in test_layers.py on linux?

Can you test your change with this configuration?

        python.version: '3.5'
        ONNX_PATH: onnx==1.2.3
        KERAS: keras==2.1.6

**Comment by vinitra** [comment]
> > Hi @jiafatom, I separated the alphabetical reordering changes away from the actual content changes. Any suggestions for how to help resolve the failing test in test_layers.py on linux?
> 
> Can you test your change with this configuration?
> 
> ```
>     python.version: '3.5'
>     ONNX_PATH: onnx==1.2.3
>     KERAS: keras==2.1.6
> ```

After talking to David, it seems like the tolerances are too high and which is causing flaky test results. Submitted an update with new recommended test tolerances (1.e-6 instead of 1.e-8).
